### PR TITLE
Grove implementation of Jordan's quantum gradient estimation algorithm

### PR DESCRIPTION
**Pull request by kmckiern**
PR includes code, unit tests, and example notebook.

The code consists of the primary code, ```jordan_gradient.py```, which uses the phase estimation technique to 'kickback' samples from an oracle function.  Also included are a set of helper functions for working with and converting between real decimals and binary fractions, ```gradient_utils.py```.

Reference
10.1103/PhysRevLett.95.050501

Originally inspired by
10.1063/1.3266959

### CODE REVIEW

**Comment by kmckiern** [comment]
I've run some additional benchmarks for the quadratic potential described in ```JordanGradient.ipynb```.

For
``` x = array([ 0.44442222,  0.79623983,  0.88588272]) ```
``` h = [0.01, 0.1, 1] ```
``` precision = [4, 6, 8, 10, 12] ```
``` n_measurements = [50, 100] ```
Estimate the gradient and compute the absolute error.

As expected, ``` h ``` > 0.1 is too large.
h

Removing ``` h ``` > 0.1, compare precision.  10 qubits appears to be ideal.
precision

Removing ``` precision ``` < 10, compare number of measurements.  It appears that the estimates converge within 50 measurements.
nm

**Comment by jotterbach** [comment]
Thanks for the PR @kmckiern. We'll get up to speed on this and will review!

**Comment by jotterbach** [comment]
@ampolloreno the tests are failing because of a recent pyquil upgrade, seems unconnected to this PR

**Comment by kmckiern** [comment]
Sounds good!  Thanks

File: grove/tests/jordan_gradient/test_jordan_gradient.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import pytest
+from grove.alpha.jordan_gradient.jordan_gradient import estimate_gradient
+from mock import patch
+import pyquil.quil as pq
+import numpy as np
```
Comment (ampolloreno) [inline grove/tests/jordan_gradient/test_jordan_gradient.py]: PEP8 import ordering, same as above.

File: grove/tests/jordan_gradient/test_gradient_utils.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import pytest
+from grove.alpha.jordan_gradient.gradient_utils import real_to_binary, \
+    binary_to_real, measurements_to_bf
+import numpy as np
```
Comment (ampolloreno) [inline grove/tests/jordan_gradient/test_gradient_utils.py]: PEP8 import ordering, same as above.

**Comment by jotterbach** [comment]
@kmckiern we fixed some underlying issues of grove that now fixes the build system again. Can you rebase your PR on top of recent master to see potential failures in the CI tool?

**Comment by kmckiern** [comment]
@ampolloreno @jotterbach thanks!  All of your suggestions sound good.  I will address them within the next few days.

File: grove/alpha/jordan_gradient/jordan_gradient.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
+from __future__ import division
```
[the remark this answers is not in the document]
Comment (kmckiern) [reply grove/alpha/jordan_gradient/jordan_gradient.py]: Attempted this in 98833b840485d958ef380e7ae12604f5f9708c7d but apparently it's not quite right. This line is a standard library import and therefore should be first, no?

[the remark this answers is not in the document]
Comment (ampolloreno) [reply grove/alpha/jordan_gradient/jordan_gradient.py]: I think that's right - I didn't mean to suggest that `from __future__` should go somewhere else, I just clicked the top of the file to comment about imports. Sorry about that!

[the remark this answers is not in the document]
Comment (kmckiern) [reply grove/alpha/jordan_gradient/jordan_gradient.py]: got it, thanks

**Comment by kmckiern** [comment]
@ampolloreno @jotterbach I think your suggests are address.  LNK if I missed anything or you can think of anything else.  Thanks!

File: grove/alpha/jordan_gradient/jordan_gradient.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    """Encode f_h into ancilla eigenvalue and kickback to input registers
+
+    :param float f_h: Oracle output at perturbation h.
+    :param int precision: Bit precision of gradient.
+    :return: Quil program to perform phase kickback.
+    :rtype: Program
+    """
+
+    # encode f_h / 2 into CPHASE gate
+    U = np.array([[np.exp(-1.0j * np.pi * f_h), 0],
```
Comment (ampolloreno) [inline grove/alpha/jordan_gradient/jordan_gradient.py]: In the paper, Jordan writes that's the phase should actually be `1.j * 2 * np.pi/m * f_h` It looks like the reason this works in the notebook is because you picked the example of 1/2k(x - .5)**2 for x\in[0, 1], so m ends up being 1. (I think) Here `f_h` should be scaled by the range of the gradient of f, so that this works for other spring constants. One way would be to have an input to the function that is the size of the interval of values the gradient can take on. For now, the user can just provide this, which the paper says is reasonable for any numerical algorithm.

I was also puzzled by the fact that the paper says 2pi and you have 1pi here, but that's because he uses a fourier transform, and you're using phase estimation. Am I right in saying that? This code certainly will calculate the gradient, as you've demonstrated, but phase estimation is more expensive than doing the QFT.

In addition, constructing this unitary might be non trivial, and doing a controlled version of this gate in the QFT even more so.

To summarize, I think we should pass a variable that specifies the size of range of values the gradient can take on, and I think we should note in the docstring for this function that we are using quantum phase estimation, rather than the quantum fourier transform, as described in the paper.

File: grove/alpha/jordan_gradient/gradient_utils.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    elif isinstance(number, float):
+        n_sign = np.sign(number)
+
+    deci = 0
+    for ndx, val in enumerate(str(number).split('.')[-1]):
+        deci += float(val) / 2**(ndx+1)
+    deci *= n_sign
+
+    return deci
+
```
Comment (ampolloreno) [inline grove/alpha/jordan_gradient/gradient_utils.py]: PEP8 Two newlines between functions.

File: grove/tests/jordan_gradient/test_jordan_gradient.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import pytest
+from mock import patch
+import pyquil.quil as pq
```
Comment (ampolloreno) [inline grove/tests/jordan_gradient/test_jordan_gradient.py]: unused import

File: grove/tests/jordan_gradient/test_jordan_gradient.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import numpy as np
+import pytest
```
Comment (ampolloreno) [inline grove/tests/jordan_gradient/test_jordan_gradient.py]: unused import

File: grove/alpha/jordan_gradient/jordan_gradient.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    :return: Quil program to initialize this system.
+    :rtype: Program
+    """
+
+    # ancilla qubit to plane wave state
+    ic_ancilla = pq.Program([X(ancilla_qubit), H(ancilla_qubit)])
+    p_ic = pq.Program(ic_ancilla)
+
+    return p_ic
+
```
Comment (ampolloreno) [inline grove/alpha/jordan_gradient/jordan_gradient.py]: PEP8 two newlines between module level functions, and can you change it for other functions in this file too! (If you use an IDE like pycharm it will whine about things like this)

File: grove/alpha/jordan_gradient/jordan_gradient.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+from grove.alpha.phaseestimation.phase_estimation import phase_estimation
+
+from grove.alpha.jordan_gradient.gradient_utils import binary_to_real, \
+    measurements_to_bf
+
+
+def initialize_system(ancilla_qubit):
+    """Prepare initial state
+
+    :param list ancilla_qubit: Qubit of ancilla register.
```
Comment (ampolloreno) [inline grove/alpha/jordan_gradient/jordan_gradient.py]: This is not a list, right?

File: grove/alpha/jordan_gradient/jordan_gradient.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
+from __future__ import division
+
+import numpy as np
+import pyquil.quil as pq
+from pyquil.gates import X, H
+from grove.alpha.phaseestimation.phase_estimation import phase_estimation
+
+from grove.alpha.jordan_gradient.gradient_utils import binary_to_real, \
```
Comment (ampolloreno) [inline grove/alpha/jordan_gradient/jordan_gradient.py]: The newline is unnecessary, there's enough space for `measurements_to_bf` on this line.

File: grove/alpha/jordan_gradient/jordan_gradient.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    # enumerate input and ancilla qubits
+    input_qubits = list(range(precision))
+    ancilla_qubit = precision
+
+    # generate gradient program
+    perturbation_sign = np.sign(f_h)
+    p_gradient = gradient_estimator(f_h, ancilla_qubit)
+
+    # run gradient program
+    if not cxn:
```
Comment (ampolloreno) [inline grove/alpha/jordan_gradient/jordan_gradient.py]: I think the more Pythonic thing to do would be default set `cxn=None` and check `if cxn is None`, and also move the import to the top of the file.

These lines are also not being tested.

File: grove/alpha/jordan_gradient/jordan_gradient.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        from pyquil.api import QVMConnection
+        cxn = QVMConnection()
+    measurements = cxn.run(p_gradient, input_qubits, n_measurements)
+    if isinstance(measurements, str):
```
Comment (ampolloreno) [inline grove/alpha/jordan_gradient/jordan_gradient.py]: This isn't being tested, and also I don't think it's necessary. You're trying to catch some atypical response from the connection, I think it would be okay to let it error, unless we just aren't handling the output from `cxn.run` correctly here.

File: examples/JordanGradient.ipynb [hunk 13]
```diff
@@ -0,0 +1,354 @@
...
+   "execution_count": 10,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "def potential_gradient(x, h, n_measurements, precision, precision_map, cxn=None):\n",
+    "    \n",
+    "    # if you are running this live, input your qvm or qpu object to cxn\n",
+    "    if not cxn:\n",
```
Comment (ampolloreno) [inline examples/JordanGradient.ipynb]: This should be `if cxn is None`, this is from PEP8: https://www.python.org/dev/peps/pep-0008/#programming-recommendations

File: grove/tests/jordan_gradient/test_gradient_utils.py [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+import pytest
+
+from grove.alpha.jordan_gradient.gradient_utils import binary_to_real, \
+    measurements_to_bf
+
+
+def test_binary_to_real():
+    precision = 9
```
Comment (ampolloreno) [inline grove/tests/jordan_gradient/test_gradient_utils.py]: `precision`  isn't being used.

File: grove/alpha/jordan_gradient/gradient_utils.py [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+    """Convert binary fraction to real decimal
+
+    :param float number: Floating point representation of binary fraction.
+    :return: Real decimal representation of binary fraction.
+    :rtype: float
+    """
+
+    if isinstance(number, str):
+        if number[0] == '-':
+            n_sign = -1
```
Comment (ampolloreno) [inline grove/alpha/jordan_gradient/gradient_utils.py]: I think this line is not being covered by tests.

**Comment by kmckiern** [comment]
@ampolloreno thanks for your feedback, will work through your suggestions in the next few days (sorry for the delay, I have been traveling)

File: grove/alpha/jordan_gradient/jordan_gradient.py (hunk 10 above)
Comment (kmckiern) [reply grove/alpha/jordan_gradient/jordan_gradient.py]: joining `measurements_to_bf` makes this line 89 characters, I think pep8 specifies a maximum of 79, no?

Comment (ampolloreno) [reply grove/alpha/jordan_gradient/jordan_gradient.py]: You're right! Sorry, my IDE was set to 100!

File: grove/alpha/jordan_gradient/jordan_gradient.py (hunk 4 above)
Comment (kmckiern) [reply grove/alpha/jordan_gradient/jordan_gradient.py]: Scaling for f_h (the m term, now variable `gradient_max`) has been added.  I also updated the example notebook to include this.

So, Jordan's algorithm employs a phase kickback, which is a form of phase estimation, and the end of the phase kickback uses a qft. The end of the `phase_estimation` function currently implemented in Grove also uses a qft.  I've reframed the code to have the 2pi term, using the Grove `phase_estimation` function. It just came down to adjusting the initialization of the ancilla register. In the process, this made the code a bit simpler.

**Comment by kmckiern** [comment]
@ampolloreno Thanks for the review, I believe I have addressed everything.  LMKWYT

**Comment by ampolloreno** [comment]
Looks great, thank you @kmckiern!

@jotterbach LGTM.

**Comment by jotterbach** [comment]
LGTM! Thanks @kmckiern for this PR, thanks @ampolloreno for the review!
